### PR TITLE
Change match desugaring to degenerate to if-else-if chains

### PR DESCRIPTION
**Pull request by nikomatsakis**
In previous PRs, I changed the match desugaring to generate more efficient code for ints/chars and the like. But this doesn't help when you're matching strings, ranges, or other crazy complex things (leading to #29740). This commit restructures match desugaring _yet again_ to handle that case better -- basically we now degenerate to an if-else-if chain in such cases.

~~Note that this builds on https://github.com/rust-lang/rust/pull/29763 which will hopefully land soon. So ignore the first few commits.~~ landed now

r? @Aatch since he's been reviewing the other commits in this series

### CODE REVIEW

**Comment by larsbergstrom** [comment]
It would be great to land this ASAP, since this bug is blocking Servo's rustup :-)

File: src/librustc_mir/build/matches/mod.rs [hunk 1]
```diff
@@ -311,17 +340,126 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
         }
 
         // If there are no candidates that still need testing, we're done.
         // Since all matches are exhaustive, execution should never reach this point.
         if candidates.is_empty() {
-            return self.panic(block);
+            return vec![block];
+        }
+
+        // Test cawndidates where possible.
```
Comment (nrc) [inline src/librustc_mir/build/matches/mod.rs]: typo

File: src/librustc_mir/build/matches/mod.rs [hunk 2]
```diff
@@ -311,17 +340,126 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
+    ///   other lists.
+    ///
+    /// So we'll compile the test. For each outcome of the test, we
+    /// recursively call `match_candidates` with the corresponding set
+    /// of candidates. But note that this set is now inexhaustive: for
+    /// example, in the case where the test returns false, there are
+    /// NO candidates, even though there is stll a value to be
+    /// matched. So we'll collect the return values from
+    /// `match_candidates`, which are the blocks where control-flow
+    /// goes if none of the candidates matched. At this point, we can continu with the "otherwise" list.
```
Comment (nrc) [inline src/librustc_mir/build/matches/mod.rs]: typo: continu, also wrapping overflow

**Comment by nrc** [comment]
r+ with the typos fixed

**Comment by Manishearth** [comment]
Fails tidy, needs https://github.com/Manishearth/rust/commit/4f655382c8f85f95d7bb05a6be46417fad42679a

**Pull request merged**